google/cloud/testing_util/opentelemetry_matchers.cc

```diff
@@ -0,0 +1,73 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+#include "google/cloud/testing_util/opentelemetry_matchers.h"
+#include "absl/memory/memory.h"
+#include <opentelemetry/sdk/instrumentationscope/instrumentation_scope.h>
+#include <opentelemetry/sdk/trace/simple_processor.h>
+#include <opentelemetry/sdk/trace/tracer.h>
+#include <opentelemetry/sdk/trace/tracer_provider_factory.h>
+#include <opentelemetry/trace/provider.h>
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace testing_util {
+
+using ::testing::AllOf;
+using ::testing::Matcher;
+using ::testing::Pointee;
+using ::testing::Property;
+
+Matcher<SpanDataPtr> SpanHasInstrumentationScope() {
+  return Pointee(
+      Property(&opentelemetry::sdk::trace::SpanData::GetInstrumentationScope,
```

Comment:
Can we use a lambda instead of `Property` and so avoid taking the address of a foreign symbol?

Reply:
http://google.github.io/googletest/gmock_cook_book.html#validating-a-member-of-an-object

The gMock Cookbook agrees: "Don’t use `Property()` against member functions that you do not own, because taking addresses of functions is fragile and generally not part of the contract of the function."

I will work on it....

Reply:
One suggestion I saw was ...

> if you currently have `Property(&Foo::bar, SubMatcher())`, you can replace it with `BarIs(SubMatcher())` where BarIs is defined by a matcher macro like
> ```
> MATCHER_P(BarIs, m) { return testing::ExplainMatchResult(m, arg.bar(), result_listener); }
> ```

Reply:
Thanks, I was kind of stuck, and your suggestion gave me some direction.

I decided not to use `::testing::ExplainMatchResult` because it would just dump the `arg` object, not `arg.bar()`. So I manually added in the relevant information. This is what an error looks like: (scroll to the end)

```
google/cloud/internal/opentelemetry_test.cc:88: Failure
Value of: spans[0]
Expected: span named (name: "darren")
  Actual: (ptr = 0x556157162fc0, value = 304-byte object <90-CB D5-55 61-55 00-00 7B-FE C7-3F 52-00 E8-C7 3F-9D 8B-71 D4-B9 09-9C A6-13 00-2F 6D-A0 76-E5 01-00 00-00 00-00 00-00 98-CF D5-55 61-55 00-00 00-1E 16-57 61-55 00-00 C0-1D 16-57 61-55 00-00 ... 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 02-00 00-00 00-00 00-00 60-35 16-57 61-55 00-00 00-27 16-57 61-55 00-00>), span name: span1)
```

Reply:
Is there any opportunity to unify the shapes of `span named (name: "darren")` and `span name: span1`?